module/metrics/rest_api.go

```diff
@@ -0,0 +1,137 @@
+	httpRequestDurHistogram   *prometheus.HistogramVec
+	httpResponseSizeHistogram *prometheus.HistogramVec
+	httpRequestsInflight      *prometheus.GaugeVec
+	httpRequestsTotal         *prometheus.GaugeVec
+}
+
+// NewRecorder returns a new metrics recorder that implements the recorder
+// using Prometheus as the backend.
+func NewRecorder(cfg metricsProm.Config) CustomRecorder {
+	if len(cfg.DurationBuckets) == 0 {
+		cfg.DurationBuckets = prometheus.DefBuckets
+	}
+
+	if len(cfg.SizeBuckets) == 0 {
+		cfg.SizeBuckets = prometheus.ExponentialBuckets(100, 10, 8)
+	}
+
+	if cfg.Registry == nil {
+		cfg.Registry = prometheus.DefaultRegisterer
+	}
+
+	if cfg.HandlerIDLabel == "" {
+		cfg.HandlerIDLabel = "handler"
+	}
+
+	if cfg.StatusCodeLabel == "" {
+		cfg.StatusCodeLabel = "code"
+	}
+
+	if cfg.MethodLabel == "" {
+		cfg.MethodLabel = "method"
+	}
+
+	if cfg.ServiceLabel == "" {
+		cfg.ServiceLabel = "service"
+	}
+
+	r := &recorder{
+		httpRequestDurHistogram: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: cfg.Prefix,
+			Subsystem: "http",
+			Name:      "request_duration_seconds",
+			Help:      "The latency of the HTTP requests.",
+			Buckets:   cfg.DurationBuckets,
+		}, []string{cfg.ServiceLabel, cfg.HandlerIDLabel, cfg.MethodLabel, cfg.StatusCodeLabel}),
+
+		httpResponseSizeHistogram: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: cfg.Prefix,
+			Subsystem: "http",
+			Name:      "response_size_bytes",
+			Help:      "The size of the HTTP responses.",
+			Buckets:   cfg.SizeBuckets,
+		}, []string{cfg.ServiceLabel, cfg.HandlerIDLabel, cfg.MethodLabel, cfg.StatusCodeLabel}),
+
+		httpRequestsInflight: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: cfg.Prefix,
+			Subsystem: "http",
+			Name:      "requests_inflight",
+			Help:      "The number of inflight requests being handled at the same time.",
+		}, []string{cfg.ServiceLabel, cfg.HandlerIDLabel}),
+
+		httpRequestsTotal: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: cfg.Prefix,
+			Subsystem: "http",
+			Name:      "requests_total",
+			Help:      "The number of requests handled over time.",
+		}, []string{cfg.ServiceLabel, cfg.HandlerIDLabel}),
+	}
+
+	return r
+}
+
+type responseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func MetricsMiddleware() mux.MiddlewareFunc {
+	r := NewRecorder(metricsProm.Config{Prefix: "access_rest_api"})
+	metricsMiddleware := middleware.New(middleware.Config{Recorder: r})
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			// This is a custom metric being called on every http request
+			r.AddTotalRequests(req.Context(), req.Method, req.URL.Path)
+
+			// Modify the writer
+			respWriter := &responseWriter{w, http.StatusOK}
+
+			// Record go-http-metrics/middleware metrics and continue to the next handler
+			std.Handler("", metricsMiddleware, next).ServeHTTP(respWriter, req)
+		})
+	}
+}
```

Comment:
from an organizational perspective, I think the `MetricsMiddleware` should be defined in `engine/access/rest` to keep it with the other HTTP server functionality

It could then accept a `RestCollector` implementation as an argument. This aligns more closely with our other uses of metrics collectors, and simplifies testing with the `NoopCollector`